src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -721,42 +727,56 @@ impl Tree {
+        // since we need to 1) do a post-traversal and 2) remove nodes from the tree.
+        // Since we do a post-traversal (by deleting nodes only after handling all children),
+        // we also need to be a bit smarter than "pop node, push all children."
+        let mut stack = vec![(root, 0)];
+        while let Some((tag, nth_child)) = stack.last_mut() {
+            let node = self.nodes.get(*tag).unwrap();
+            if *nth_child == node.children.len() {
+                // We have visited all children. Let's remove those that are useless.
+                // Split the children into useful and useless.
+                // FIXME: this copies new_children, even if nothing gets removed.
+                // A 2-pass approach could be more efficient. It could consist of
+                // 1. traverse the Tree, collect all useless tags in a Vec
+                // 2. traverse the Vec, remove all tags previously selected
+                let (useless, remaining_children) =
+                    node.children.iter().copied().partition(|x| self.is_useless(*x, live));
+                // Assign the new children.
+                // This might make this node useless as well, but removal is the parent's job.
+                self.nodes.get_mut(*tag).unwrap().children = remaining_children;
+                // Next, ensure all useless children are properly unregistered everywhere.
+                for idx in useless {
+                    // Note: In the rest of this comment, "this node" means that at idx.
+                    // This node has no more children (if there were any, they have already been removed).
+                    // It is also unreachable as determined by the GC, so we can remove it everywhere.
+                    // Due to the API of UniMap we must absolutely call
+                    // `UniValMap::remove` for the key of this tag on *all* maps that used it
+                    // (which are `self.nodes` and every range of `self.rperms`)
+                    // before we can safely apply `UniKeyMap::remove` to truly remove
+                    // the tag from the `tag_mapping`.
+                    let node = self.nodes.remove_owned(idx).unwrap();
+                    for (_perms_range, perms) in self.rperms.iter_mut_all() {
+                        perms.remove(idx);
+                    }
+                    self.tag_mapping.remove(&node.tag);
+                }
+                // We are done, the parent can continue.
+                stack.pop();
+                continue;
+            } else {
+                // We are starting to visit that subtree.
+                // For that, we push a new frame with the initial conditions, and remember that we continue with the next once that subtree is done.
+                let next_child = node.children[*nth_child];
+                *nth_child += 1;
+                stack.push((next_child, 0));
```

Comment:
```suggestion
                // Visit the child by pushing it to the stack.
                // Also increase `nth_child` so that when we come back to the `tag` node, we
                // look at the next child.
                let next_child = node.children[*nth_child];
                *nth_child += 1;
                stack.push((next_child, 0));
```